src/body/response/body.rs

```diff
@@ -0,0 +1,225 @@
+use bytes::{Bytes, BytesMut};
+use futures_core::Stream;
+use hyper::body::{Body, Frame, SizeHint};
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use super::{Buffered, Either, Mapped, Streaming};
+use crate::{Error, Result};
+
+pub struct ResponseBody {
+    body: Either<Either<Buffered, Streaming>, Box<Mapped>>,
+}
+
+/// A projection type for the possible variants of the nested `Either` enums
+/// that compose the `body` field of the `ResponseBody` struct.
+//
+// This `enum` primarily exists to get a `Pin<&mut Mapped>` reference to the
+// `Box<Mapped>` contained in the `Either::Right` variant of the `body` field
+// of the `ResponseBody` struct so we can delegate the call to `poll_frame` to
+// `Mapped` struct when necessary since it is `!Unpin`.
+enum ResponseBodyProjection<'a> {
+    /// The projection type for a `Buffered` response body.
+    Buffered(Pin<&'a mut Buffered>),
+    /// The projection type for a `Mapped` response body.
+    Mapped(Pin<&'a mut Mapped>),
+    /// The projection type for a `Streaming` response body.
+    Streaming(Pin<&'a mut Streaming>),
+}
+
+impl ResponseBody {
+    pub fn is_empty(&self) -> bool {
+        self.len().map_or(false, |len| len == 0)
+    }
+
+    pub fn len(&self) -> Option<usize> {
+        match &self.body {
+            Either::Left(Either::Left(buffered)) => Some(buffered.len()),
+            Either::Right(mapped) => mapped.len(),
+            _ => None,
+        }
+    }
+}
+
+impl ResponseBody {
+    pub(crate) fn new() -> Self {
+        Self::buffer(Bytes::new())
+    }
+
+    pub(crate) fn buffer(data: Bytes) -> Self {
+        let buffer = Box::new(BytesMut::from(data));
+        let buffered = Buffered::new(buffer);
+
+        Self {
+            body: Either::Left(Either::Left(buffered)),
+        }
+    }
+
+    pub(crate) fn stream<T>(stream: T) -> Self
+    where
+        T: Stream<Item = Result<Bytes>> + Send + 'static,
+    {
+        let stream = Streaming::new(Box::new(stream));
+
+        Self {
+            body: Either::Left(Either::Right(stream)),
+        }
+    }
+
+    pub(crate) fn map<F>(self, map: Box<F>) -> Self
```

Comment:
Here's a revised version of your PR comment with improved grammar:

---

I think we may want to create a facade struct that contains the `body` field of `ResponseBody` but does not implement the `body` trait. This struct could later be wrapped in another struct that does implement `Body`. This approach would ensure that it is impossible to move data out of the `Pin<&mut Self>` in the `poll_frame` method. While it's not as straightforward as using `map` to `&mut self`, it would give me peace of mind knowing that a mistake that could otherwise go unnoticed would result in a compiler error. Perhaps we can construct the pollable wrapper around `ResponseBody` by mapping the `inner` response before it is returned from `Response::into_inner`.

Reply:
I think my concerns have been addressed by [2e9575c](https://github.com/zacharygolba/via/pull/9/commits/2e9575cba112765e3eb91f10858189d9f1f595cf). I'm not going to resolve this conversation in the event that anyone has thoughts or feedback about this approach.